src/index/LocatedTriples.cpp

```diff
@@ -0,0 +1,240 @@
+// Copyright 2023 - 2024, University of Freiburg
+// Chair of Algorithms and Data Structures
+// Authors:
+//    2023 Hannah Bast <bast@cs.uni-freiburg.de>
+//    2024 Julian Mundhahs <mundhahj@tf.uni-freiburg.de>
+
+#include "index/LocatedTriples.h"
+
+#include <algorithm>
+
+#include "absl/strings/str_join.h"
+#include "index/CompressedRelation.h"
+#include "index/IndexMetaData.h"
+#include "index/Permutation.h"
+
+IdTriple permute(const IdTriple& triple,
+                 const std::array<size_t, 3>& keyOrder) {
+  return {triple[keyOrder[0]], triple[keyOrder[1]], triple[keyOrder[2]]};
+}
+
+std::vector<LocatedTriple> LocatedTriple::locateTriplesInPermutation(
+    const std::vector<IdTriple>& triples, const Permutation& permutation,
+    bool shouldExist) {
+  const Permutation::MetaData& meta = permutation.metaData();
+  const vector<CompressedBlockMetadata>& blocks = meta.blockData();
+
+  vector<LocatedTriple> out;
+  out.reserve(triples.size());
+  size_t currentBlockIndex;
+  for (auto triple : triples) {
+    triple = permute(triple, permutation.keyOrder());
+    currentBlockIndex =
+        std::ranges::lower_bound(
+            blocks, triple,
+            [&](const IdTriple& block, const IdTriple& triple) {
+              return block < triple;
+            },
+            [](const CompressedBlockMetadata& block) {
+              const auto& perm = block.lastTriple_;
+              return IdTriple{perm.col0Id_, perm.col1Id_, perm.col2Id_};
+            }) -
+        blocks.begin();
+    out.emplace_back(currentBlockIndex, triple, shouldExist);
+  }
+
+  return out;
+}
+
+// ____________________________________________________________________________
+std::pair<size_t, size_t> LocatedTriplesPerBlock::numTriples(
+    size_t blockIndex) const {
+  // If no located triples for `blockIndex_` exist, there is no entry in `map_`.
+  if (!map_.contains(blockIndex)) {
+    return {0, 0};
+  }
+
+  auto blockUpdateTriples = map_.at(blockIndex);
+  size_t countDeletes = std::ranges::count_if(
+      blockUpdateTriples,
+      [](const LocatedTriple& elem) { return !elem.shouldTripleExist_; });
+  size_t countInserts = blockUpdateTriples.size() - countDeletes;
+
+  return {countInserts, countDeletes};
+}
+
+size_t LocatedTriplesPerBlock::mergeTriples(size_t blockIndex, IdTable block,
+                                            IdTable& result,
+                                            size_t offsetInResult,
+                                            size_t numIndexColumns) const {
+  // This method should only be called if there are located triples in the
+  // specified block.
+  AD_CONTRACT_CHECK(map_.contains(blockIndex));
+
+  // TODO<qup42>: We're assuming that the index columns are always {0, 1, 2},
+  // {1, 2} or {2}. Is this true?
```

Comment:
That is a good point. I will add that.

Reply:
This remark was originally only targeted at the index columns (=the components of the triple). What I had in mind was something like `CompressedRelationReader::getDistinctColIdsAndCountsImpl`. At first look I suspect that [this](https://github.com/ad-freiburg/qlever/blob/master/src/index/CompressedRelation.cpp#L471-L474) could also only give `{1}`as the index columns.

Reply:
I have been playing testing some ways to add the payload. Once used, will the payload always be the some or could it be different for different triples?

Reply:
1. Currently All triples have the same load given a permutation. But the payloads may differ per permutation (more payloads for PSO/POS for example.

2. I just figured out: The merging only works if you have ALL THREE columns, because otherwise everything breaks. 
    So for your mergeRoutine, you alsways need the full columns, only afterwards you may get a smaller subset of columns.